### ISSUE TITLE
Detecting stable branch and channel does not work without CONAN_CHANNEL environment variable

### ISSUE THREAD

**Issue by Artalus** [conan-io/conan-package-tools#481]
In accordance with #479 , if you use CPT to build a Linux package with reference like `lib/ver@/`, it will not upload built packages even if the stable branch+channel detection should've been triggered. E.g. in [this build](https://github.com/trassir/conan-center-index/runs/488017980?check_suite_focus=true#step:5:145) for [this commit](https://github.com/trassir/conan-center-index/commit/5e87a2b19a620d834bd366bb9485118dae82457c#diff-2955d5257f635de1df53f55a171ca5c7R13) CPT in Linux build says `  >> Skipping upload, not stable channel`. Contrary, in [Windows build](https://github.com/trassir/conan-center-index/runs/488018001?check_suite_focus=true#step:5:110) it says `>> Redefined channel by CI branch matching with 'trassir-ci', setting CONAN_CHANNEL to '_'`.

I did some digging and the channel selection strategy looks rather weird to me.
- First, in [`ConanMultiPackager` constructor](https://github.com/conan-io/conan-package-tools/blob/417cefb3c590d8fe7b40e0543cc4128be0239108/cpt/packager.py#L206) we call `_get_specified_channel()` with `channel` passed to us from parameters, and `reference` from either parameters or envvars.
- In [said function](https://github.com/conan-io/conan-package-tools/blob/417cefb3c590d8fe7b40e0543cc4128be0239108/cpt/packager.py#L794) we may attempt to get reference from envvar again for some reason.
Then we check if the reference contains `@` - and this check only decides on what the fallback channel parameter should be, `'testing'` or `None`. So this function doesn't even try to parse the channel from reference string, and will happily resolve to `None` if the envvar is not set (due to #479 it cannot be set on Linux without causing build fail).
Then we call `_get_channel()` with selected channel string (possibly `None`).
- In [`_get_channel`](https://github.com/conan-io/conan-package-tools/blob/417cefb3c590d8fe7b40e0543cc4128be0239108/cpt/packager.py#L764), however, we immediately return if the channel string is empty, without attempting to check for current branch name and possibly switch said empty string to something else.
- Finally, [later in the constrcutor](https://github.com/conan-io/conan-package-tools/blob/417cefb3c590d8fe7b40e0543cc4128be0239108/cpt/packager.py#L316) we for some reason setup the `CONAN_CHANNEL` envvar if the selected channel string is not empty.

Ignoring the channel not being parsed from reference, it seems that removing the not-none check in `_get_channel()` should fix the problem (and it does in my local fork). However, given the quite fragile process of setting up the builder, not yet fixed #479, and the possible removal of `CONAN_CHANNEL` envvar in #407, I am unsure if said change won't actually break something else.

### Environment Details
* Conan Package Tools Version: **0.31.1**
* Operating System: **Linux**
* Conan version: **conan 1.22.2**

**Comment by uilianries** [comment]
Hi @Artalus ! Thanks for this detailed explanation.

**Comment by uilianries** [comment]
I'm curious, why do you need package without user/channel if only CCI is using it for official packages? Also, did you try set CONAN_CHANNEL=_ ? Regards.

**Comment by Artalus** [comment]
Mostly because many official packages from CCI use dependencies in said style: e.g. if we require `openssl/1.1.1c` it will require `zlib/1.2.11` in its `conanfile.py` without user/channel specified. If we were to build everything as `lib/ver@company/stable`, we'd have to edit dependencies in official recipes, which we'd like to omit if possible. However we need binaries of packages from CCI built for toolchains that are not used in upstream CCI - which is why the question of rebuilding official packages arises.

We cannot use `CONAN_CHANNEL` since on Linux it causes build errors from #479 . Though over the holiday I realized that the stuff I describe in this issue here might be kinda unrelated to CPT - as in specifying reference as `lib/ver@/` instead of `lib/ver` might be a bug in itself, rather than a feature, so I'd better devote my attention to #479 for now.

**Comment by Hopobcn** [comment]
Yep, I also observed this. Very annoying to be honest. If you decide you want a `lib/ver@` style of reference, together with `upload_only_when_stable=True` you'll for sure hit this issue.

The alternative of setting `CONAN_CHANNEL=_`, `CONAN_USERNAME=_` & `
CONAN_STABLE_CHANNEL=_` doesn't work. Because all builds will be considered stable disregarding the `stable_branch_pattern`.
